transaction/issuance.go

```diff
@@ -0,0 +1,126 @@
+package transaction
+
+import (
+	"encoding/json"
+	"errors"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/vulpemventures/fastsha256"
+	"github.com/vulpemventures/go-elements/confidential"
+	"github.com/vulpemventures/go-elements/internal/bufferutil"
+)
+
+// IssuanceEntity defines one of the fields of the issuance contract
+type IssuanceEntity struct {
+	Domain string `json:"domain"`
+}
+
+// IssuanceContract defines the structure of the Ricardian contract of the issuance
+type IssuanceContract struct {
+	Name      string         `json:"name"`
+	Ticker    string         `json:"ticker"`
+	Version   uint           `json:"version"`
+	Precision uint           `json:"precision"`
+	Entity    IssuanceEntity `json:"entity"`
+}
+
+// TxIssuance defines the type for Issuance field in TxInput
+type TxIssuance struct {
+	AssetBlindingNonce []byte
+	AssetEntropy       []byte
+	AssetAmount        []byte
+	TokenAmount        []byte
+}
+
+// NewTxIssuance returns a new empty issuance instance
+func NewTxIssuance(assetAmount, tokenAmount uint64) (*TxIssuance, error) {
+	if assetAmount < 0 {
+		return nil, errors.New("invalid asset amount")
+	}
+	if tokenAmount < 0 {
+		return nil, errors.New("invalid token amount")
+	}
+
+	aa, err := confidential.SatoshiToElementsValue(assetAmount)
+	if err != nil {
+		return nil, err
+	}
+	ta, err := confidential.SatoshiToElementsValue(tokenAmount)
+	if err != nil {
+		return nil, err
+	}
+
+	return &TxIssuance{
+		AssetAmount: aa[:],
+		TokenAmount: ta[:],
+	}, nil
+}
+
+// GenerateEntropy generates the entropy from which the hash of the asset and
+// of the reissuance token are calculated
+func (issuance *TxIssuance) GenerateEntropy(inTxHash []byte, inTxIndex uint32, contract *IssuanceContract) error {
+	if len(inTxHash) != 32 {
+		return errors.New("invalid tx hash length")
+	}
+
```

Comment:
we should check the provided contract matches the asset amount provided

Reply:
the contract does not have information about the amount of the asset

Reply:
the contract no, but the TxIssuance struct yes.